### PR TITLE
Feature/weaviate memory

### PR DESCRIPTION
**Pull request by cs0lar**
### Background

This change allows the use of different vector-based memories for Auto-GPT as long as the `Memory` interface is implemented. As an example, this change integrates [Weaviate](https://weaviate.io/) as an alternative to Pinecone that can be instantiated locally.

### Changes

- We created a `providers` directory which is meant to hold the various memory implementations. 
- The original `scripts/memory.py` file has been moved into the `providers` module and it declares `Memory` as an interface to be implemented.
- A new `scripts/factory.py` file has been added which declares a factory method for creating the appropriate memory support class. A new environment variable, `MEMORY_PROVIDER` allows to configure which provider to use.
- The Pinecone implementation has been moved into the `providers` module.
- An Weaviate implementation of the `Memory` interface has been added.

### Test Plan

Added factory tests in `tests/memory_tests.py`

### Change Safety

- [x] I have added tests to cover my changes
- [ ] I have considered potential risks and mitigations for my changes

### CODE REVIEW

**Comment by cs0lar** [comment]
I have noticed that the latest commands list in commands.py no longer supports the memory-related command (`memory_add`, memory_del`, `memory_ovr`). In this pull request I have removed the associated handlers (commit_memory, delete_memory and overwrite_memory) in part because they no longer follow the memory interface. Let me know whether these memory commands should be back on the menu and I can implement them.

**Review by Bakuutin — APPROVED** [review]

**Comment by dahifi** [comment]
Thank you so much I was actually working on a similar branch last week! 😃

**Review by nponeccop — APPROVED** [review]

**Comment by cs0lar** [comment]
I have now merged the memory strategy from the master branch and added weaviate to the supported providers. I have also added tests for the weaviate integration.

File: README.md [hunk 1]
```diff
@@ -185,13 +185,15 @@ Pinecone enables the storage of vast amounts of vector-based memory, allowing fo
 2. Choose the `Starter` plan to avoid being charged.
 3. Find your API key and region under the default project in the left sidebar.
 
+
```
Comment (nponeccop) [inline README.md]: Revert the unrelated whitespace changes

File: README.md [hunk 2]
```diff
@@ -185,13 +185,15 @@ Pinecone enables the storage of vast amounts of vector-based memory, allowing fo
...
 ### Setting up environment variables
 
 Simply set them in the `.env` file. 
 
 Alternatively, you can set them from the command line (advanced):
 
 For Windows Users:
+
```
Comment (nponeccop) [inline README.md]: Another one

File: scripts/commands.py [hunk 3]
```diff
@@ -298,4 +298,4 @@ def delete_agent(key):
     result = agents.delete_agent(key)
     if not result:
         return f"Agent {key} does not exist."
-    return f"Agent {key} deleted."
+    return f"Agent {key} deleted."
```
Comment (nponeccop) [inline scripts/commands.py]: This is optional, but add an extra line feed at the end of file to remove this (-) sign from the diff. I will still approve if you won't

File: scripts/main.py [hunk 4]
```diff
@@ -415,4 +415,3 @@ def parse_arguments():
             chat.create_chat_message(
                 "system", "Unable to execute command"))
         print_to_console("SYSTEM: ", Fore.YELLOW, "Unable to execute command")
-
```
Comment (nponeccop) [inline scripts/main.py]: Revert this whitespace change

File: scripts/memory/pinecone.py [hunk 5]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (nponeccop) [inline scripts/memory/pinecone.py]: And this

File: scripts/memory/pinecone.py [hunk 6]
```diff
@@ -48,4 +47,4 @@ def get_relevant(self, data, num_relevant=5):
         return [str(item['metadata']["raw_text"]) for item in sorted_results]
 
     def get_stats(self):
-        return self.index.describe_index_stats()
+        return self.index.describe_index_stats()
```
Comment (nponeccop) [inline scripts/memory/pinecone.py]: Here you removed the final line feed. Just revert

File: scripts/memory/weaviate.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            print(f'Unexpected error {err=}, {type(err)=}')
+            return []
+
+    def get_stats(self):
+        result = self.client.query.aggregate(self.index) \
+                     .with_meta_count() \
+                     .do()
+        class_data = result['data']['Aggregate'][self.index]
+
+        return class_data[0]['meta'] if class_data else {}
```
Comment (nponeccop) [inline scripts/memory/weaviate.py]: You can add a line feed here as well (optional)

**Review by nponeccop — CHANGES_REQUESTED** [review]
LGTM but needs whitespace fixes.

**Review by nponeccop — DISMISSED** [review]

**Comment by hsm207** [comment]
@cs0lar what are your thoughts on adding support for [embedded weaviate](https://weaviate.io/developers/weaviate/installation/embedded) too as part of this PR?

**Comment by cs0lar** [comment]
@hsm207 Thanks for spotting that. I have added weaviate embedded support now. Please check it out.

File: .env.template [hunk 8]
```diff
@@ -7,9 +7,22 @@ FAST_LLM_MODEL=gpt-3.5-turbo
...
+WEAVIATE_EMBEDDED_PATH="/home/me/.local/share/weaviate"
+WEAVIATE_USERNAME=
+WEAVIATE_PASSWORD=
+IMAGE_PROVIDER=dalle
+HUGGINGFACE_API_TOKEN=
 OPENAI_AZURE_API_BASE=your-base-url-for-azure
 OPENAI_AZURE_API_VERSION=api-version-for-azure
 OPENAI_AZURE_DEPLOYMENT_ID=deployment-id-for-azure
 IMAGE_PROVIDER=dalle
-HUGGINGFACE_API_TOKEN=
```
Comment (nponeccop) [inline .env.template]: Undo the relocation of the HUGGINGFACE line

**Review by nponeccop — CHANGES_REQUESTED** [review]

File: .env.template (hunk 8 above)
Comment (cs0lar) [reply .env.template]: thanks! there were also some dupes I have now removed.

**Review by nponeccop — DISMISSED** [review]

File: README.md [hunk 9]
```diff
@@ -204,6 +204,22 @@ export PINECONE_ENV="Your pinecone region" # something like: us-east4-gcp
 
 ```
 
+## Weaviate Setup
+
+[Weaviate](https://weaviate.io/) is an open-source vector database. It allows to store data objects and vector embeddings from ML-models and scales seamlessly to billion of data objects. [An instance of Weaviate can be created locally (using Docker), on Kubernetes or using Weaviate Cloud Services](https://weaviate.io/developers/weaviate/quickstart).
```
Comment (hsm207) [inline README.md:359]: @cs0lar should also mention embedded weaviate here.

Comment (cs0lar) [reply README.md:359]: good spot, thanks! This is now done.

File: README.md [hunk 10]
```diff
@@ -204,6 +204,26 @@ export PINECONE_ENV="Your pinecone region" # something like: us-east4-gcp
 
 ```
 
+## Weaviate Setup
+
+[Weaviate](https://weaviate.io/) is an open-source vector database. It allows to store data objects and vector embeddings from ML-models and scales seamlessly to billion of data objects. [An instance of Weaviate can be created locally (using Docker), on Kubernetes or using Weaviate Cloud Services](https://weaviate.io/developers/weaviate/quickstart). 
+Although still experimental, [Embedded Weaviate](https://weaviate.io/developers/weaviate/installation/embedded) is supported which allows the Auto-GPT process itself to start a Weaviate instance. To enable it, set `USE_WEAVIATE_EMBEDDED` to `True`. 
+
+#### Setting up enviornment variables
```
Comment (hsm207) [inline README.md]: Followed the instructions and ... it did not work 😂

Because it is missing the weaviate client.

So, perhaps edit to say something like "First, run `pip install weaviate-client`?

Comment (cs0lar) [reply README.md]: @hsm207  HA!  I had assumed `pip install -r requirements.txt` would have been run. But you raise a good point. Have added to README.

**Review by nponeccop — APPROVED** [review]

**Comment by hsm207** [comment]
@cs0lar question about the schema:

```
{
    "class": "Autogpt",

    },
    "properties": [
        {
            "dataType": [
                "text"
            ],
            "description": "original text for the embedding",
            "name": "raw_text",
            "tokenization": "word"
        },
        {
            "dataType": [
                "text"
            ],
            "description": "This property was generated by Weaviate\'s auto-schema feature on Wed Apr 12 16:53:47 2023",
            "name": "class",
            "tokenization": "word"
        }
    ],
   ...
}'
```
is the property `class` intended? Asking because it was autogenerated.

File: scripts/memory/weaviate.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+                "name": "raw_text",
+                "dataType": ["text"],
+                "description": "original text for the embedding"
+            }
+        ],
+    }
+
+class WeaviateMemory(MemoryProviderSingleton):
+    def __init__(self, cfg):
+        auth_credentials = self._build_auth_credentials(cfg)
```
Comment (hsm207) [inline scripts/memory/weaviate.py]: @cs0lar I think adding support for API key would be immensely useful for users getting started using weaviate's free sandbox environment, so they can have a similar experience with pinecone i.e. just provide an api key and url:

https://weaviate.io/developers/weaviate/client-libraries/python#api-key-authentication

File: README.md [hunk 12]
```diff
@@ -204,6 +204,26 @@ export PINECONE_ENV="Your pinecone region" # something like: us-east4-gcp
 
 ```
 
+## Weaviate Setup
+
+[Weaviate](https://weaviate.io/) is an open-source vector database. It allows to store data objects and vector embeddings from ML-models and scales seamlessly to billion of data objects. [An instance of Weaviate can be created locally (using Docker), on Kubernetes or using Weaviate Cloud Services](https://weaviate.io/developers/weaviate/quickstart). 
+Although still experimental, [Embedded Weaviate](https://weaviate.io/developers/weaviate/installation/embedded) is supported which allows the Auto-GPT process itself to start a Weaviate instance. To enable it, set `USE_WEAVIATE_EMBEDDED` to `True` and make sure you `pip install "weaviate-client>=3.15.4`.
```
Comment (aslafy-z) [inline README.md]:
```suggestion
Although still experimental, [Embedded Weaviate](https://weaviate.io/developers/weaviate/installation/embedded) is supported which allows the Auto-GPT process itself to start a Weaviate instance. To enable it, set `USE_WEAVIATE_EMBEDDED` to `True` and make sure you `pip install "weaviate-client>=3.15.4"`.
```

File: scripts/memory/weaviate.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+        with self.client.batch as batch:
+            batch.add_data_object(
+                uuid=doc_uuid,
+                data_object=data_object,
+                class_name=self.index,
+                vector=vector
+            )
+
+            batch.flush()
```
Comment (hsm207) [inline scripts/memory/weaviate.py]: @cs0lar this call is unnecessary since you are using the batch context manager

**Comment by hsm207** [comment]
> @hsm207 Thanks for spotting that. I have added weaviate embedded support now. Please check it out.

@cs0lar it's not clear to me how the auto gpt user should configure the .env file if they want to use embedded weaviate.

File: scripts/memory/weaviate.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    def add(self, data):
+        vector = get_ada_embedding(data)
+
+        doc_uuid = generate_uuid5(data, self.index)
+        data_object = {
+            'class': self.index,
+            'raw_text': data
+        }
+
+        with self.client.batch as batch:
```
Comment (hsm207) [inline scripts/memory/weaviate.py]: @cs0lar is data always going to be uploaded one at a time from auto gpt to weaviate?

File: README.md [hunk 15]
```diff
@@ -204,6 +204,26 @@ export PINECONE_ENV="Your pinecone region" # something like: us-east4-gcp
...
+
+#### Setting up enviornment variables
+
+In your `.env` file set the following:
+
+```
+MEMORY_BACKEND=weaviate
+WEAVIATE_HOST="127.0.0.1" # the IP or domain of the running Weaviate instance
+WEAVIATE_PORT="8080" 
+WEAVIATE_PROTOCOL="http"
```
Comment (hsm207) [inline README.md:370]: We should add something like "set `USE_WEAVIATE_EMBEDDED=True` if you want to use embedded weaviate"

**Comment by nponeccop** [comment]
@cs0lar There are conflicts again due to the massive merge

**Review by nponeccop — APPROVED** [review]

File: scripts/memory/weaviate.py (hunk 11 above)
Comment (cs0lar) [reply scripts/memory/weaviate.py]: added

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@cs0lar LGTM, but there are conflicts now due to merging of other PRs.

**Review by nponeccop — APPROVED** [review]

File: scripts/memory/weaviate.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        self._create_schema()
+
+    def _create_schema(self):
+        schema = default_schema(self.index)
+        if not self.client.schema.contains(schema):
+            self.client.schema.create_class(schema)
+
+    def _build_auth_credentials(self, cfg):
+        if cfg.weaviate_username and cfg.weaviate_password:
+            return weaviate_auth.AuthClientPassword(cfg.weaviate_username, cfg.weaviate_password)
```
Comment (hsm207) [inline scripts/memory/weaviate.py]: @cs0lar this line will throw an error as `weaviate_auth` is not defined anymore in this module

**Comment by nponeccop** [comment]
@cs0lar There are conflicts now though.

File: scripts/memory/weaviate.py (hunk 16 above)
Comment (hsm207) [reply scripts/memory/weaviate.py]:
```suggestion
            return weaviate.AuthClientPassword(cfg.weaviate_username, cfg.weaviate_password)
```@cs0lar this line will throw an error as `weaviate_auth` is not defined anymore in this module

**Review by nponeccop — APPROVED** [review]

**Review by nponeccop — DISMISSED** [review]

File: scripts/memory/weaviate.py (hunk 16 above)
Comment (hsm207) [reply scripts/memory/weaviate.py]: @cs0lar I think you missed this change

File: .env.template [hunk 17]
```diff
@@ -57,6 +57,7 @@ SMART_TOKEN_LIMIT=8000
 ################################################################################
 
 # MEMORY_BACKEND - Memory backend type (Default: local)
+>>>>>>> master
```
Comment (nponeccop) [inline .env.template]: This isn't right

**Review by nponeccop — CHANGES_REQUESTED** [review]

**Comment by nponeccop** [comment]
@cs0lar There are conflicts now

**Comment by bobvanluijt** [comment]
Hi, @nponeccop and @cs0lar - is there a way how we can coordinate the fix and merge? Otherwise, we might keep going in circles :)

**Comment by cs0lar** [comment]
@bobvanluijt it would be great to know what's holding merging this branch (apart from the latest conflicts). If this functionality is not part of the plan it would be great to know sometime soon so we can move on from it. BTW honour to @nponeccop for relentlessly checking the PRs and conflict fixes!

File: .env.template (hunk 17 above)
Comment (cs0lar) [reply .env.template]: fixed thanks

**Review by nponeccop — DISMISSED** [review]

**Comment by richbeales** [comment]
Run flake8 autogpt/ tests/ --select E303,W293,W291,W292,E305,E231,E302
autogpt/memory/__init__.py:30:1: E302 expected 2 blank lines, found 1
autogpt/memory/__init__.py:56:1: W293 blank line contains whitespace
autogpt/memory/base.py:10:1: E302 expected 2 blank lines, found 1
autogpt/memory/weaviate.py:9:1: E302 expected 2 blank lines, found 1
autogpt/memory/weaviate.py:21:1: E302 expected 2 blank lines, found 1
autogpt/memory/weaviate.py:76:5: E303 too many blank lines (2)
autogpt/memory/weaviate.py:80:80: W291 trailing whitespace
tests/integration/weaviate_memory_tests.py:14:1: E302 expected 2 blank lines, found 1
tests/integration/weaviate_memory_tests.py:41:1: W293 blank line contains whitespace
tests/integration/weaviate_memory_tests.py:47:1: W293 blank line contains whitespace
tests/integration/weaviate_memory_tests.py:56:1: W293 blank line contains whitespace
tests/integration/weaviate_memory_tests.py:70:1: W293 blank line contains whitespace
tests/integration/weaviate_memory_tests.py:87:5: E303 too many blank lines (2)
tests/integration/weaviate_memory_tests.py:102:5: E303 too many blank lines (2)

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@cs0lar There are conflicts again

**Comment by bobvanluijt** [comment]
> @bobvanluijt it would be great to know what's holding merging this branch (apart from the latest conflicts). 

Thanks, @nponeccop, I think it's the merging of other PRs. Seems we are almost there. Your help is appreciated in keeping this PR up-to-date 🙏

Almost there, it seems: https://github.com/Significant-Gravitas/Auto-GPT/pull/424#issuecomment-1509865221

> BTW honour to @nponeccop for relentlessly checking the PRs and conflict fixes!

Yup, thanks @nponeccop

**Review by nponeccop — DISMISSED** [review]

**Comment by bobvanluijt** [comment]
Awesome! I think the PR can be merged 🥳

**Review by hsm207 — CHANGES_REQUESTED** [review]

**Review by hsm207 — DISMISSED** [review]

**Comment by bobvanluijt** [comment]
@cs0lar, I think we need your help again. People on the Auto-GPT Discord are asking for this merge 😂

**Comment by bobvanluijt** [comment]
Hi @nponeccop and @richbeales. I think the merge conflict is resolved. Would be awesome if you could review and merge. Thanks 🙏

**Comment by richbeales** [comment]
Thanks for reaching out @bobvanluijt - we'll get to this shortly.  I'll ping you when ready.

**Review by BillSchumacher — APPROVED** [review]

**Review by richbeales — APPROVED** [review]

**Review by BillSchumacher — APPROVED** [review]

**Comment by level20peon** [comment]
I think Milvus and Weaviate should also be appended here, right? https://github.com/Significant-Gravitas/Auto-GPT/blob/master/.env.template#L53

**Comment by bobvanluijt** [comment]
🎉 thanks all for the help and merge 🎉 

Added a minor change here: https://github.com/Significant-Gravitas/Auto-GPT/pull/1813

Basically, so that people don't have to install the Weaviate client manually.
